### ISSUE TITLE
AQL: serious flaw: failure to prune paths

### ISSUE THREAD

**Issue by pkoppstein** [arangodb/arangodb#3979]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [✅] 3.2
- [ ] self-compiled devel branch

Mode:
- [ ] Cluster
- [✅] Single-Server

Storage-Engine:
- [✅]] mmfiles
- [ ] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [✅ ] MacOS, version: High Sierra

### this is an AQL-related issue:
[✅  ] I'm using graph features

I'm issuing AQL via:
- [✅  ] web interface with this browser:     running on this OS:
- [ ] arangosh
- [ ] this Driver:

Unfortunately ArangoDB v3.2.9 fails to optimize a simple traversal of the following trivial graph, G:

G consists of a large number of nodes (tasks), numbered 0, 1, ....; there are two edges from each n, both pointing to n+1: one edge has type == "before" and the other has type "start_before".

As expected, the query:
```
FOR v, e, p IN 1..100 OUTBOUND 'tasks/1' depends
 FILTER p.edges[*].type ALL == "before"
 COLLECT WITH COUNT into counter
 RETURN counter
```
runs quickly (about 1.7 seconds on my machine) and produces the correct answer.  The same results, of course, apply to the query following the "start_before" paths.  "Explain" explains that 4
optimizations are applied.

The bad news is that the variant query shown below is not optimized and in fact fails to complete in a reasonable amount of time except for small MAX values:
```
FOR v, e, p IN 1..100 OUTBOUND 'tasks/1' depends
 FILTER p.edges[*].type ALL == e.type
 COLLECT WITH COUNT into counter
 RETURN counter
```
In the present context, this query simply asks: "how many paths of the two types are there?"
and should take approximately twice as long as the fixed-RHS query (i.e. about 3.4 seconds).
"Explain" indicates that NO optimizations are applied.

The problem evidently is that the Optimizer does not realize that even though the RHS of the "ALL" condition is path-dependent, pruning should nevertheless take place, because once an edge is encountered that breaks the condition, adding another edge won't fix things.

### User-defined pruning rules

This leads to a larger question: how can one write an AQL query
to force pruning to take place -- something along the lines of:

    PRUNE IF <condition>

That way, we could write:
```
FOR v, e, p IN 1..100 OUTBOUND 'tasks/1' depends
 PRUNE IF length(p) > 1 AND p.edges[-2].type != e.type
 COLLECT WITH COUNT into counter
 RETURN counter
```

**Comment by mchacki** [comment]
Hi,

thanks for reporting.
Indeed an option to PRUNE is on our Roadmap but not yet introduced.
So in the future it will be available.

Next is your second Query cannot do any ALL optimization thus for because for technical reasons we need an "immutable" RHS of the condition for each traversal (which may be improved later)
As `e` is potentially changing in every step the optimizer has no chance for optimization.
As a quick fix you could do the following right now:

    FOR tmpV, tmpE  IN 1..1 OUTBOUND 'tasks/1' depends
      FOR v, e, p  IN 0..99 OUTBOUND tmpV depends
        FILTER p.edges[*].type ALL == tmpE.type
        COLLECT WITH COUNT into counter
        RETURN counter

So we do a single step traversal to extract a constant type.
And than we execute a traversal starting form this step that allows for proper optimization.
However i will keep this open as "FEATURE REQUEST"

**Comment by pkoppstein** [comment]
Thanks for the prompt response and for the good news about the Roadmap. I was already aware of the workaround, but only after I realized what the problem was. In retrospect, I should have paid more attention to the Explainer.

So I think there are at least three separate ERs here:

1. documentation about this specific limitation of the current Optimizer;
2. greater prominence about the upcoming PRUNE enhancement (for example in the AQL manual itself), with further details as they become available;
3. enhancement of the optimizer so that it can recognize that once an ALL condition on a full-length path has failed, pruning can take place, provided the values of the expressions on the LHS and RHS at the point of failure are fixed (which I believe amounts to saying, that both are deterministic functions of the nodes and edges of the graph).

**Comment by maxkernbach** [comment]
Hi @pkoppstein,

I am closing this issue as fixed since the PRUNE feature was introduced in 3.4.5 and 3.5 RC1.

>Pruning is the easiest variant to formulate conditions to reduce the amount of data to be checked during a search. So it allows to improve query performance and reduces the amount of overhead generated by the query. Pruning can be executed on the vertex, the edge and the path and any variable defined before.

This [video tutorial](https://www.youtube.com/watch?v=wex3Nb2RYQE) shows the difference between FILTER and the new PRUNE with a hands-on example. You can find more details in our [documentation](https://www.arangodb.com/docs/3.4/aql/graphs-traversals.html#pruning).
